Bug 1642554 - Implement model for engineer traction formula

ionutgoldan opened the pull request:
**Bug link:** https://bugzilla.mozilla.org/show_bug.cgi?id=1642554
More details there.

airimovici commented on treeherder/perf/exceptions.py:
```diff
@@ -12,3 +12,11 @@ class MissingRecords(Exception):
+    pass
+
+
+class BugzillaEndpointProblem(RuntimeError):
```
This does not sound that well, it can be renamed to `BugzillaEndpointError` or something else

ionutgoldan commented on treeherder/perf/sheriffing_criteria.py:
```diff
@@ -0,0 +1,148 @@
+from treeherder.perf.exceptions import NoFiledBugs, BugzillaEndpointError
+
+
+class EngineerTractionFormula:
```
Custom code that's talking to a web service.

ionutgoldan commented on tests/perf/test_engineer_traction.py:
```diff
@@ -0,0 +1,283 @@
+        _ = engineer_traction.breakdown()
+
+
+# Leveraging HTTP VCR
```
Test code making use of betamax & betamax-serializers *(from here down to bottom of module)*

ionutgoldan commented on tests/sample_data/betamax_cassettes/perf_sheriffing_criteria/awsy-JS.json:
```diff
@@ -0,0 +1,533 @@
+{
```
Sample betamax-generated fixture

ionutgoldan commented on tests/perf/test_engineer_traction.py:
```diff
@@ -0,0 +1,283 @@
+#  bugs that are at least 2 weeks old
+
+
+@pytest.fixture
+def session() -> Session:
+    return create_bugzilla_session()
+
+
+@pytest.fixture
+def betamax_recorder(session):
+    Betamax.register_serializer(pretty_json.PrettyJSONSerializer)
+    return Betamax(session, cassette_library_dir=CASSETTE_LIBRARY_DIR)
```
Setting up betamax for our tests

ionutgoldan commented on tests/perf/test_engineer_traction.py:
```diff
@@ -0,0 +1,283 @@
+from treeherder.utils.http import create_bugzilla_session
+
+CASSETTE_LIBRARY_DIR = 'tests/sample_data/betamax_cassettes/perf_sheriffing_criteria'
+VCR_RECORDING_DATE = 'June 2nd, 2020'
+
+
+pytestmark = [pytest.mark.freeze_time(VCR_RECORDING_DATE, tick=True)]
```
Fast backwarded time on the date we performed our VCR

ionutgoldan commented on treeherder/perf/sheriffing_criteria.py:
```diff
@@ -0,0 +1,148 @@
+    def _fetch_quantified_bugs(self, framework: str, suite: str, test: str = None) -> List[dict]:
+        test_moniker = ' '.join(filter(None, (suite, test)))
+        test_id_fragments = filter(None, [framework, test_moniker])
+        creation_time = datetime.strftime(self.oldest_timestamp, BZ_DATETIME_FORMAT)
```
Custom code that’s making time-based requests to a web service

ionutgoldan commented on tests/perf/test_engineer_traction.py:
```diff
@@ -0,0 +1,283 @@
+        return bug_resp.json()['bugs']
+
+
+@pytest.fixture
+def cooled_down_bugs(session, quantified_bugs) -> List[dict]:
+    bugs = []
+    for bug in quantified_bugs:
+        created_at = datetime.strptime(bug['creation_time'], BZ_DATETIME_FORMAT)
+        if created_at <= datetime.now() - timedelta(weeks=2):
+            bugs.append(bug)
+    return bugs
```
Test that requires more control over time-based APIs

ionutgoldan commented on tests/perf/test_engineer_traction.py:
```diff
@@ -0,0 +1,283 @@
+    assert engineer_traction.oldest_timestamp >= no_older_than
+
+
+@pytest.mark.parametrize(
+    'cooled_down_bug',
+    [
+        {"creation_time": "2020-05-18T15:20:55Z"},  # older than 2 weeks
+        {"creation_time": "2020-05-04T15:20:55Z"},  # older than 1 month
+        {"creation_time": "2019-12-16T08:10:37Z"},  # older than 6 months
+    ],
+)
+def test_formula_correctly_detects_cooled_down_bugs(cooled_down_bug, session):
+    engineer_traction = EngineerTractionFormula(session)
+
+    assert engineer_traction.has_cooled_down(cooled_down_bug)
```
Test that require more control over time-based APIs

ionutgoldan commented on tests/perf/test_engineer_traction.py:
```diff
@@ -0,0 +1,283 @@
+    assert engineer_traction.has_cooled_down(cooled_down_bug)
+
+
+@pytest.mark.parametrize(
+    'not_cooled_down_bug',
+    [
+        {'creation_time': '2020-05-31T00:00:00Z'},  # 2 days old
+        {'creation_time': '2020-05-26T00:00:00Z'},  # 1 week old
+        {'creation_time': '2020-05-19T23:00:00Z'},  # ~2 weeks old, except for 1 hour
+    ],
+)
+def test_formula_detects_bugs_that_didnt_cool_down_yet(not_cooled_down_bug, session):
+    engineer_traction = EngineerTractionFormula(session)
+
+    assert not engineer_traction.has_cooled_down(not_cooled_down_bug)
```
Test that require more control over time-based APIs

ionutgoldan commented on tests/perf/test_engineer_traction.py:
```diff
@@ -0,0 +1,283 @@
+# Leveraging HTTP VCR
+
+
+def test_breakdown_updates_between_calculations(betamax_recorder):
+    engineer_traction = EngineerTractionFormula(betamax_recorder.session)
+
+    test_moniker_A = ('build_metrics', 'build times')
+    test_moniker_B = ('talos', 'tp5n', 'nonmain_startup_fileio')
+
+    cassette_preffix_A = '-'.join(filter(None, test_moniker_A))
+    cassette_preffix_B = '-'.join(filter(None, test_moniker_B))
+
+    with betamax_recorder.use_cassette(f'{cassette_preffix_A}', serialize_with='prettyjson'):
+        engineer_traction(*test_moniker_A)  # let it perform calculus & cache breakdown
+        breakdown_A = engineer_traction.breakdown()
+
+    with betamax_recorder.use_cassette(f'{cassette_preffix_B}', serialize_with='prettyjson'):
+        engineer_traction(*test_moniker_B)  # let it perform calculus & cache breakdown
+        breakdown_B = engineer_traction.breakdown()
+
+    assert breakdown_A != breakdown_B
```
Test that require more control over time-based APIs *(VCR is assuming we're always on June 2nd, 2020)*

klahnakoski replied:
@ionutgoldan where is the artifact that tells VCR it is June 2nd, 2020?

ionutgoldan replied:
I've setup this global var, in the test module:
```python
CASSETTE_LIBRARY_DIR = 'tests/sample_data/betamax_cassettes/perf_sheriffing_criteria'
```
So I suggest we store all recordings in `tests/sample_data/betamax_cassettes` & organize them there in sub dirs, depending on the feature/component/module that's tested.

ionutgoldan commented on requirements/dev.in:
```diff
@@ -10,6 +10,10 @@ PyPOM
 # for git commit hooks
 pre-commit
 
+# for test driven development
+pytest-testmon
+pytest-watch
```
Test runners optimized for test driven development

klahnakoski replied:
@ionutgoldan  Please add the `ptw --runner “pytest --testmon”` command to the `tests/README.md` in the correct spot

ionutgoldan commented:
> 
> 
> I'm only reviewing adding the dependencies and the .gitignore.
> It looks good from my point of you.
> 
> I encourage you to add a bit of documentation on how the tools you're adding can be used to generate fixutres and may prove useful to other engineers to understand the benefits.

@armenzg @camd @sarah-clements @klahnakoski I've sent you an email about the testing tools I proposed in this PR. I've also shared with you a detailed Google Doc, explaining more on them + provided some comments in this PR with the relevant code sections that are leveraging the tools.

We can start discussing about them here.

ionutgoldan commented:
> This PR has test enhancements detailed https://docs.google.com/document/d/1a4gISVkCHTd-YIvVUZtuAWXdKGt0YBlPWb_noIWcfUM/edit# **PLUS** code for `EngineerTrackingFormula`. Yes? Nothing else?

That's correct. Nothing else.
The PR enhancements are accompanied by the VCR fixtures that were collected automatically. That's the main reason why this PR is more than 4,000 lines of code.

ionutgoldan commented:
>     * Is there no database schema change, or did I miss it?

Yep, no database schema change. This is just business logic that's querying Bugzilla.

>     * How many calls does this make to Bugzilla in practice?

We plan to run this every 6 months. When this happens, there will be *lots* of calls to Bugzilla.
To give a very rough estimation, I can say this will make around 1,000 demanding calls (can last several tens of seconds), followed by 1,000 * 15 light-weight calls (no more than 2 seconds each, maybe less).

>     * Is this `EngineerTrackingFormula` a stateless code atop Bugzilla calls?
>     * Where is the VCR storing the test time? is it per-test?  Must test/code use a special time function to use this feature?

The VCR is not storing the test time, even though its requests are time dependent.
I'm using `pytest-freezegun` to control the "system" time. If configured it at module level with this construct:
```python
VCR_RECORDING_DATE = 'June 2nd, 2020'

pytestmark = [pytest.mark.freeze_time(VCR_RECORDING_DATE, tick=True)]
```

ionutgoldan commented:
> [...]
> By adding more headers, Bugzilla OPS will be more likely to contact you before blocking your IP when you run this.

Good point. I'm going to update the code to address this.

> We can work on using the ES-cluster-with-bug-snapshots; which was meant for scanning large numbers of bugs for reporting like this.

@klahnakoski you mean we should schedule another task to refactor `EngineerTractionFormula`? So that it's communicating with ES-cluster-with-bug-snapshots instead of Bugzilla?

> I like the `requests` recordings (I wish we had that on the frontend side (do we already?).

Glad you do. I've been thinking about the frontend side, too. But I haven't dig through to find that out? @camd do we have a similar Javascript-based VCR tool?

> Being able to lock the time is promising; I hope it works (since the database can also be a source of time).

👍 

> I do not believe `pytest-watch` & `pytest-testmon` will be useful to me, but they do not get in my way. If they are useful to you then it is good.

I'm not too keen on integrating the `pytest-watch` & `pytest-testmon` combo in this PR. I'll actually remove them from dev dependencies, so this PR is more focused on the time & recording test tools.

I'll file a dedicated PR, in which I'll provide some detailed markdown docs. My main goal is to reduce the test session time from 10 minutes to less than 1min, using the necessary techniques.

ionutgoldan commented:
@klahnakoski I've addressed the change requests related to the testing tools.

Briefly, I've:
* reverted my changes over the `commenter.py` & `http.py` modules, as those weren't needed anymore.
* derived a special special Session class, which guarantees the mandatory `Referer` & `User-agent` headers; enforced `EngineerTractionFormula` to work only with it
* tweaked the associated tests (some renames, covered the new changes for `EngineerTractionFormula`)

If all looks good from this perspective, we can continue reviewing the bug's implementation itself.

klahnakoski commented on tests/README.md:
```diff
@@ -56,13 +56,27 @@ Django can perform a number of checks to ensure you are configured correctly
+
     pytest tests
 
+### Start TDD-enhanced test runner in the background
+
+Whenever code changes are detected within the project, command bellow will run only the associated tests.
+Test runs will be triggered for either application or tests code changes.
+
+This workflow ensures a considerably faster feedback loop for those wanting to do test driven development.
+
+    ptw --runner "pytest --testmon" --clear
+
```
Thank you!

klahnakoski commented on treeherder/perf/sheriffing_criteria.py:
```diff
@@ -8,16 +10,47 @@
+class NonBlockableSession(Session):
+    def __init__(self, referer=None):
+        super().__init__()
+        referer = referer or PERF_SHERIFFING_CRITERIA
```
I like the reference to the google doc!

klahnakoski commented on treeherder/intermittents_commenter/commenter.py:
```diff
@@ -210,7 +210,15 @@ def update_whiteboard(self, existing, new):
+            'x-bugzilla-api-key': settings.COMMENTER_API_KEY,
+            'Accept': 'application/json',
+        }
+        return session
```
I agree with the in-lining you did here.
